### PR TITLE
Fixes for content assist with parameterized rule calls

### PR DESCRIPTION
**Pull request by szarnekow**
This PR extracts a couple of internal / anonymous classes to top-level types to make the code easier to understand and customize. Also it fixes NoSuchMethod errors when invoking CA with complex grammars that use Parameterized Rules. Last but not least the entry configuration for lookahead tokens wrt to the applied rule call arguments is fixed.

### CODE REVIEW

**Comment by szarnekow** [comment]
Tests to come.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * A wrapper for an existing {@link IUnorderedGroupHelper} that ignores the first
+ * invocation of {@link #enter(UnorderedGroup)}.
+ * 
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * @since 2.13
+ */
+public class IgnoreFirstEntranceUnorderedGroupHelper implements IUnorderedGroupHelper {
+	private final IUnorderedGroupHelper helper;
+	boolean first = true;
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java]: package private by intention?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+	/**
+	 * Protected contructor to allow to override this type. 
+	 * 
+	 * @see BaseContentAssistParser#ignoreFirstEntrance(IUnorderedGroupHelper)
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java]: helper may not be null?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/LookAheadBasedTokenSource.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+/*******************************************************************************
+ * Copyright (c) 2017 TypeFox GmbH (http://www.typefox.io) and others.
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/LookAheadBasedTokenSource.java]: itemis copyright?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java (hunk 1 above)
Comment (szarnekow) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java]: Fixed

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java (hunk 2 above)
Comment (szarnekow) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java]: Added Preconditions check

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/LookAheadBasedTokenSource.java (hunk 3 above)
Comment (szarnekow) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/LookAheadBasedTokenSource.java]: Fixed

**Comment by kthoms** [comment]
Could you add `@since 2.13` to new public/protected methods, please?

**Comment by szarnekow** [comment]
Looks like I will need to add @since 2.14

**Comment by cdietrich** [comment]
yes and you should rebase the eclipse branch (and possible this one too)

**Comment by kthoms** [comment]
It is not too late for 2.13 yet

**Comment by szarnekow** [comment]
It's a bigger change, not tested yet and not fully covered by unit tests. I don't see this for 2.13

**Comment by cdietrich** [comment]
@szarnekow anything left to keep this unmerged?

**Comment by szarnekow** [comment]
@cdietrich yes, it still need more testing and refinement. I'll come back to this soon.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/BaseContentAssistParser.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+
+/**
+ * Abstract base type for the {@link AbstractContentAssistParser} and the deprecated
+ * equivalent in the xtext.ui bundle.
+ * 
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * @since 2.13
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/BaseContentAssistParser.java]: 2.14

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/EofListener.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import org.antlr.runtime.RecognizerSharedState;
+import org.eclipse.xtext.AbstractElement;
+import org.eclipse.xtext.UnorderedGroup;
+import org.eclipse.xtext.ide.editor.contentassist.antlr.ObservableXtextTokenStream.StreamListener;
+import org.eclipse.xtext.ide.editor.contentassist.antlr.internal.AbstractInternalContentAssistParser;
+import org.eclipse.xtext.parser.antlr.IUnorderedGroupHelper;
+
+/**
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * @since 2.13
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/EofListener.java]: 2.14

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import org.eclipse.xtext.parser.antlr.IUnorderedGroupHelper;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * A wrapper for an existing {@link IUnorderedGroupHelper} that ignores the first
+ * invocation of {@link #enter(UnorderedGroup)}.
+ * 
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * @since 2.13
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/IgnoreFirstEntranceUnorderedGroupHelper.java]: 2.14

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/LookAheadBasedTokenSource.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+package org.eclipse.xtext.ide.editor.contentassist.antlr;
+
+import java.util.Iterator;
+
+import org.antlr.runtime.Token;
+import org.antlr.runtime.TokenSource;
+
+/**
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * @since 2.13
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/antlr/LookAheadBasedTokenSource.java]: 2.14

**Pull request merged**